sticking in pouchdb

wangfowen opened the pull request:
I'm not too well-versed in latest trends in javascript. did I do promises / callbacks correctly?

wangfowen committed:
sticking in pouchdb

sherbondy reviewed (commented):
There's no need to have a top level "recipes" document stored in pouchdb.

You can instead use `bulkDocs` and `allDocs` to create/query the list of all recipes.

Having the data model this way should resolve the issue where loading an individual recipe right now will have metadata missing (e.g. the recipe name, etc) unless you have already loaded the "recipes" list. 

It should also simplify bulk insertion of recipes when the peachworks response is paginated: no need for merging/appending to some parent "recipes" document.

Sorry I haven't been a great collaborator since returning from vacation, attention is being pulled in a lot of different directions right now. Will try and carve out a big chunk of time this evening to work on this.

sherbondy committed:
prettier all the things

- Add pre-commit hook to auto-prettify and fix eslint issues via lint-staged
- Update some older dependencies, e.g. eslint and node sass to play nicely with new tooling
- Remove doubly-listed dependencies from devDependencies

sherbondy committed:
merge recipe data and use allDocs.

- This demonstrates using allDocs to fetch the recipe list
- Unfortunately, we are still doing slow individual inserts for the recipe list items
- to allow us to upsert and merge with individual recipe endpoint data.
- Eventually, when using a worker job to populate the database, we should actually
  be able to generate an entirely separate database instance and then discard the old
  database altogether, so we could hypothetically do a single insert of all recipes
  in one go...
- We are also using dangerouslySetInnerHTML to actually render the recipe instructions
  html now... It is "trusted content"...

sherbondy commented on server/db.js:
```diff
@@ -9,51 +10,63 @@ let db = new pouchDb("peachworks", { adapter: "memory" });
 //endpoint which fetches from db for recipe data
 const getRecipe = (req, res) => {
-  const id = req.params.id;
+  const id = parseInt(req.params.id, 10);
```
Hmm, I probably should be more defensive about this, since you could type any (non-integer) gibberish in the url field and have it be passed into the route.

sherbondy commented:
Okay, added some automated tooling with pre-commit hook to format code and eslint fix.

Also demonstrating doing a batch query to pull all of the recipes, and having the recipe document data merged.

I guess the trendy new thing in js land is to use async/await.

It is in the stable version of node, so we could certainly use if promises have got you bummed out.

wangfowen commented on server/peachworks.js:
```diff
@@ -1,70 +1,75 @@
 
   let resultQuery = Object.assign({}, otherQueries, {
-    "access_token": peachworksAccessToken
+    access_token: peachworksAccessToken
```
I wanted to make this one a string to make it obvious it needs to remain underscored instead of camel case for the api to work. does it break the linting?

wangfowen commented on server/peachworks.js:
```diff
@@ -1,70 +1,75 @@
-    '/' + url + '?' + resultQs
+    "https://api.peachworks.com/v1/accounts/" +
+    peachworksAccountId +
+    "/" +
```
ehh do you actually find this more readable? seems a bit unnecessary to me

wangfowen reviewed (commented):
Mm feeling okay about promises, we can switch to async/await if it's better? What's the appeal there?
I just think I'm doing something wrong since I'm doing a mix of callback and promises. I wasn't sure how to chain the one promise within the promise (I made the outer promise, inner one is pouchdb created), since returning on the inner one gives undefined and returning the outer one doesn't wait for the inner one to finish, so I did a callback to chain the inner promise.

wangfowen commented on server/db.js:
```diff
@@ -9,51 +10,63 @@ let db = new pouchDb("peachworks", { adapter: "memory" });
-    .then(_doc => {
-      doc._rev = _doc._rev;
-      db.put(doc);
+    .upsert(newDoc._id, storedDoc => {
```
ah! they did implement upsert! I saw a mention of it when I looked up create or update, but didn't see it in the pouchdb api docs

wangfowen replied:
ah nvm, I see it's a plugin

wangfowen commented on server/db.js:
```diff
@@ -9,51 +10,63 @@ let db = new pouchDb("peachworks", { adapter: "memory" });
+    const recipes = recipesJson.json.results;
+    const idRecipes = recipes.map(recipe => {
+      const recipeId = "recipe-" + recipe.id;
+      return Object.assign({}, recipe, { _id: recipeId });
```
are we able to do the object splat here also? `{...recipe, _id: recipeId}`

wangfowen commented on server/db.js:
```diff
@@ -9,51 +10,63 @@ let db = new pouchDb("peachworks", { adapter: "memory" });
+    });
+
+    const noop = () => {};
+    const promises = idRecipes.map(recipe => createOrUpdate(recipe, noop));
```
if we're going to have no op callbacks, should make it a default in the createOrUpdate method instead. also, why map through all of them and createOrUpdate? why not bulkDocs?

wangfowen replied:
nvm about the bulkDocs -- just saw your commit comment. it's because we can't bulk upsert
